cpp/benchmarks/interop/interop.cpp

```diff
@@ -0,0 +1,258 @@
+                                           cudf::type_id::TIMESTAMP_SECONDS,
+                                           cudf::type_id::TIMESTAMP_MILLISECONDS,
+                                           cudf::type_id::TIMESTAMP_MICROSECONDS,
+                                           cudf::type_id::TIMESTAMP_NANOSECONDS,
+                                           cudf::type_id::DURATION_SECONDS,
+                                           cudf::type_id::DURATION_MILLISECONDS,
+                                           cudf::type_id::DURATION_MICROSECONDS,
+                                           cudf::type_id::DURATION_NANOSECONDS,
+                                           cudf::type_id::STRING,
+                                           cudf::type_id::LIST,
+                                           cudf::type_id::DECIMAL32,
+                                           cudf::type_id::DECIMAL64,
+                                           cudf::type_id::DECIMAL128,
+                                           cudf::type_id::STRUCT>;
+
+static char const* stringify_type(cudf::type_id value)
+{
+  switch (value) {
+    case cudf::type_id::INT8: return "INT8";
+    case cudf::type_id::INT16: return "INT16";
+    case cudf::type_id::INT32: return "INT32";
+    case cudf::type_id::INT64: return "INT64";
+    case cudf::type_id::UINT8: return "UINT8";
+    case cudf::type_id::UINT16: return "UINT16";
+    case cudf::type_id::UINT32: return "UINT32";
+    case cudf::type_id::UINT64: return "UINT64";
+    case cudf::type_id::FLOAT32: return "FLOAT32";
+    case cudf::type_id::FLOAT64: return "FLOAT64";
+    case cudf::type_id::BOOL8: return "BOOL8";
+    case cudf::type_id::TIMESTAMP_DAYS: return "TIMESTAMP_DAYS";
+    case cudf::type_id::TIMESTAMP_SECONDS: return "TIMESTAMP_SECONDS";
+    case cudf::type_id::TIMESTAMP_MILLISECONDS: return "TIMESTAMP_MILLISECONDS";
+    case cudf::type_id::TIMESTAMP_MICROSECONDS: return "TIMESTAMP_MICROSECONDS";
+    case cudf::type_id::TIMESTAMP_NANOSECONDS: return "TIMESTAMP_NANOSECONDS";
+    case cudf::type_id::DURATION_DAYS: return "DURATION_DAYS";
+    case cudf::type_id::DURATION_SECONDS: return "DURATION_SECONDS";
+    case cudf::type_id::DURATION_MILLISECONDS: return "DURATION_MILLISECONDS";
+    case cudf::type_id::DURATION_MICROSECONDS: return "DURATION_MICROSECONDS";
+    case cudf::type_id::DURATION_NANOSECONDS: return "DURATION_NANOSECONDS";
+    case cudf::type_id::DICTIONARY32: return "DICTIONARY32";
+    case cudf::type_id::STRING: return "STRING";
+    case cudf::type_id::LIST: return "LIST";
+    case cudf::type_id::DECIMAL32: return "DECIMAL32";
+    case cudf::type_id::DECIMAL64: return "DECIMAL64";
+    case cudf::type_id::DECIMAL128: return "DECIMAL128";
+    case cudf::type_id::STRUCT: return "STRUCT";
+    default: return "unknown";
+  }
+}
+
+NVBENCH_DECLARE_ENUM_TYPE_STRINGS(cudf::type_id, stringify_type, stringify_type)
```

Comment:
This might be used by other benchmarks too.
Please consider moving these to benchmark common header file or common utilities.

Reply:
yeah. which header would you suggest that be moved to? same header as type_id?

Reply:
I think that is a good idea. Many times, I had to implement this function for my debugging. Let's make a follow up PR, moving this into `types.hpp` (same header as `type_id`).

Reply:
we have a similar file here: https://github.com/rapidsai/cudf/blob/branch-24.12/cpp/benchmarks/io/nvbench_helpers.hpp
Worth considering a new file in https://github.com/rapidsai/cudf/tree/branch-24.12/cpp/benchmarks/common.

Declaring an NVBench-specific utility in a non-cuDF namespace within `types.hpp` is inappropriate, as it also forces `types.hpp` to include an NVBench header, which is something we want to avoid.

Reply:
I agree with a new header in the benchmarks/common. 
We could refactor some of the elements from `io/nvbench_helpers.hpp` with the new header as well.
Though the refactor may be too much for this PR.

Reply:
> agree with a new header in the benchmarks/common.

That utility function will be used not just in benchmark but also testing/debugging. I would argue against having it just in the benchmark module. Probably somewhere in `cudf/utilities` module is a better place for it.

Reply:
> That utility function will be used not just in benchmark but also testing/debugging.

Could you elaborate on this further? How does this feature specifically benefit unit tests?


Benchmarks and unit tests are treated as "external" users of libcudf, which is why they are not included within the `cudf::` namespace. The `cudf/utilities` directory is reserved for use cases within libcudf or those under the `cudf::` namespace.

Reply:
I usually debug cudf APIs by printing out the type id of the input column. For example:
```
printf("input type: %s\n", type_to_string(input.type().id()));
```

However, most of the time, for fast iteration with small issues, I just printed the type id integer value, and look up the enum, trying to identify the enum type from value. For example:
```
printf("input type: %d\n",(int)input.type().id());
```

As such, having such `type_to_string` utility is very helpful. This debug print statement can be inserted anywhere. Most of the time I insert it into the middle of a cudf API. Sometimes I use it in unit tests too. It would be most beneficial to add it into libcudf's utilities module. We can either use it for debugging libcudf APIs or for external usage in benchmark/unit tests.

Reply:
I see what you mean. It's a different feature request.

You are referring to the fact that `stringify_type` can be a common utility for both benchmarks and tests. A host-device print utility like `print_type_id` together with `stringify_type` can be added to https://github.com/rapidsai/cudf/blob/branch-24.12/cpp/include/cudf_test/print_utilities.cuh. Once it's done, this file can simply include that print utility header and add a single line for nvbench readability enhancement:
```cpp
NVBENCH_DECLARE_ENUM_TYPE_STRINGS(cudf::type_id, cudf::test::print::stringify_type, stringify_type)
```
I’m inclined to leave this refactoring for a separate PR.

Reply:
Opened #17376 to track this issue.

Reply:
Thanks for filing the issue :+1: